src/Runner/PoolRunner.php

```diff
@@ -20,220 +19,84 @@ class PoolRunner
+                });
             }
 
-            if (!$this->executeTestStep($executeCallback, $test, $pool, true)) {
-                return false;
-            }
-        }
+            // No more test wait for all remaining run
+            yield $promises;
 
-        return true;
+            Loop::stop();
+            ob_end_flush();
+        });
     }
 
-    /**
-     * Execute a test step.
-     *
-     * @param callable $callback
-     * @param Test     $test
-     * @param Pool     $pool
-     *
-     * @return bool
-     */
-    protected function executeTestStep($callback, Test $test, Pool $pool, $isTestMethod = false)
+    protected function run(Test $test): Promise
     {
-        try {
-            Assertion::$currentTest = $test;
-
-            if ($isTestMethod && $test->getMethod()->returnsReference()) {
-                $result = &$callback();
-            } else {
-                $result = $callback();
-            }
-
-            $futureHttpCollection = $this->futureHttpPool->flush();
-            $test->mergeFutureHttp($futureHttpCollection, $test);
-            $pool->queueFutureHttp($futureHttpCollection);
-        } catch (\Throwable $exception) {
-            $debugOutput = ob_get_contents();
-            ob_clean();
-
-            $this->futureHttpPool->flush();
-            $pool->passFinishTest($test);
-            $this->output->outputFailure($test, $debugOutput, $exception);
-
-            return false;
-        } catch (\Exception $exception) {
+        return \Amp\call(function () use($test) {
             $debugOutput = ob_get_contents();
             ob_clean();
 
-            $this->futureHttpPool->flush();
-            $pool->passFinishTest($test);
-            $this->output->outputFailure($test, $debugOutput, $exception);
-
-            return false;
-        }
-
-        $debugOutput = ob_get_contents();
-        ob_clean();
-
-        if ($isTestMethod) {
-            foreach ($test->getChildren() as $childTest) {
-                $childTest->addArgument($result, $test);
-            }
-        }
+            $this->output->outputStep($test, $debugOutput);
+            $test->setState(Test::STATE_RUNNING);
 
-        if ($pool->hasTest($test) && $test->getFutureHttpPool()->isEmpty()) {
-            $pool->passFinishTest($test);
-            $this->output->outputSuccess($test, $debugOutput);
+            $testCase = $this->getTestObject($test);
+            $testCase->initialize();
 
-            foreach ($test->getChildren() as $childTest) {
-                $complete = true;
+            $method = $test->getMethod()->getName();
+            $args = $test->getArguments();
 
-                foreach ($childTest->getParents() as $parentTest) {
-                    if (!$pool->hasCompletedTest($parentTest)) {
-                        $complete = false;
-                        break;
-                    }
-                }
+            try {
+                Assertion::$currentTest = $test;
```

Comment:
If they run concurrently, this won't work. But you can have global state if you decorate the event loop and save and restore the global context every time.